content/en/blog/2022-08-05-effective-programmers-read-code.md

```diff
@@ -0,0 +1,59 @@
+
+For me, the glass castle is both a structure and a map.[^2] I hold individual named components in mental space such that related components are nearer to each other. The components are linked with mental threads based on their interactions: e.g. calling a method or a function of another component or instantiating and referencing an object. When I mentally zoom into a component I see that it is itself made up of components with interactions to each other. As I code, I mentally zoom in and out of the model getting more detail when I need it and encapsulating and isolating detail when I don't.
+
+The primary purpose of the glass castle is to ensure that we can quickly locate related code in the codebase and understand the consequences of making changes to the code. It does serve another purpose however: like [code "smells"](https://martinfowler.com/bliki/CodeSmell.html), I use the glass castle to understand when there is a structural or architectural problem with the system. For example, when I zoom into a component and I see that there are too many external interactions with subcomponents, it's an indication that the parent component is not correctly abstracted. This is very similar to the idea of ["wide vs deep" modules](https://nakabonne.dev/posts/depth-of-module/) described in Ousterhout's [_A Philosophy of Software Design_](https://web.stanford.edu/~ouster/cgi-bin/book.php). The best use of the glass castle is to determine when the system is "tipping" over as we add features or make changes because it may lead us to refactoring or addressing the initial design that wasn't quite right for the current iteration of software.
+
+The glass castle is essential for how I write good code and recently, I've started to pay more attention to how I construct the glass castle. When I sit down to implement a feature or make a change, the very first thing I do is identify candidate locations in the code where the change needs to be made. These locations are identified in a top down fashion looking at the organization of the code and quickly scanning through package and module names and package documentation.
+
+Once I find the candidate locations, I continue in a bottom up fashion identifying the functional dependencies surrounding the location and putting together components and interactions tagged with the input and output data contracts that they make. Often, I have to read the code of the external dependencies of the project to figure out the best possible solutions. I continue grouping, linking, and encapsulating code in my head until I feel like I have enough of a picture that I'm confident in the change I'm about to make.
```

Comment:
Suggestions: "...I continue in a bottom-up fashion by identifying the functional dependencies..."